### ISSUE TITLE
Seemingly wrong Return(func) behavior

### ISSUE THREAD

**Issue by abourget** [vektra/mockery#71]
Such an interface:

```
type Requester interface {
	Get(path string) (string, error)
}
```

generates:

```
func (_m *Requester) Get(path string) (string, error) {
	ret := _m.Called(path)

	var r0 string
	if rf, ok := ret.Get(0).(func(string) string); ok {
		r0 = rf(path)
	} else {
		r0 = ret.Get(0).(string)
	}

	var r1 error
	if rf, ok := ret.Get(1).(func(string) error); ok {
		r1 = rf(path)
	} else {
		r1 = ret.Error(1)
	}

	return r0, r1
}
```

requiring you to use this syntax to use value generating functions:

```
type Nice interface {
   Get(string) (*Profile, error)
}

myNiceMock.On("Get").Return(func(a string) *Profile {
    if a == "hello" {
        return &Profile{}
    }
}, func(a string) error {
    if a == "hello" {
        return nil
    }
})
```

Because each return value requires its own generating function (just discovered this after reading the source a couple of times!).

Wouldn't it make more sense to have the first method be a *single* function that returns all the required parameters, instead of requiring the user to define two functions ?

What about doing something like:

```
func (_m *Requester) Get(path string) (string, error) {
	ret := _m.Called(path)

	var r0 string
	if rf, ok := ret.Get(0).(func(string) (string, error)); ok {
		return rf(path)
	} else {
		r0 = ret.Get(0).(string)
	}

	var r1 error
	r1 = ret.Error(1)

	return r0, r1
}
```

thus requiring a simpler form to use:

```
myNiceMock.On("Get").Return(func(a string) (*Profile, error) {
    if a == "hello" {
        return &Profile{}, nil
    }
})
```

Taking two functions, one for each return value, seems counterintuitive to me.

**Comment by evanphx** [comment]
I don't think it's totally counterintuitive as it allows users to have separate control of each of the return values independently.

**Comment by abourget** [comment]
It's not ?  I did violate all my expectations I must say..  and the error message I got led me to think this feature was actually not implemented at all.

For those interested in a solution: https://github.com/vektra/mockery/pull/72

**Comment by ernesto-jimenez** [comment]
@abourget you can do the following:

```myNiceMock.On("Get", "hello").Return(&Profile{}, nil)```

Wouldn't that be enough?

**Comment by abourget** [comment]
My goal was to using generative functions.. I know they are frown upon (perhaps some good reasoning about that would turn me off of implementing such a feature), but in certain situations, I can see their use.

That line wouldn't generate things based on incoming arguments..

**Comment by ernesto-jimenez** [comment]
Do you have a more clear example of what you are trying to accomplish?

It's always easer to provide feedback based on the use case instead of giving feedback on how the person is trying to address the use case :)

**Comment by abourget** [comment]
The function I'm mocking is:

```
ReceiveEvent(id string) (*Event, error)
```

and the generated `Event` needed to contain the value of `id`... having multiple concurrent consumers calling this, I thought it'd be easier by directly injecting a mock function (like I'd done in Python or JS in the past) that would use the same signature as what I was mocking, and would implement the testing behavior I wanted.

I understand the thing better now. I know how I could implement the use cases I have using the current design. It took me a lot of time though, I thought I'd avoid for the next user..

In the end, I actually backed off from using generative functions.. I'm going to use `.Once()` instead, and line up the values I need statically.. as I was able to determine order of calls with `.After()`.
